### ISSUE TITLE
Serious lag with >30 nodes

### ISSUE THREAD

**Issue by Stumblinbear** [newcat/baklavajs#38]
I can only imagine VueJS is attempting to redraw the entire scene when moving a node around, because serious lag starts to occur on my machine at around 30 nodes. Only the node in question and the connections on it should be moved/redrawn, however I'm pretty sure the entire graph is.

**Comment by newcat** [comment]
Yeah, you are right, all connections are redrawn. The reason for that is, that I am using a deep watcher on the connection. The connection contains references to both the connected node interfaces, which in turn reference their parent node, which references... Anyway, apparently everything has a reference to everything, therefore the watcher triggers pretty often and also, for example, when other nodes are moved. This can be avoided by only watching the `position` properties of the connected nodes; however, this means that height changes of the node are not detected. This can happen when an interface with an option is connected/disconnected and therefore the option is hidden/displayed.

![baklava1](https://user-images.githubusercontent.com/1523157/62073791-b624e380-b241-11e9-9b4b-12de5bb7da09.PNG)
![baklava2](https://user-images.githubusercontent.com/1523157/62073793-b624e380-b241-11e9-9187-e44865ab8b2a.PNG)

Maybe this can be done by using `ResizeObservers` but they have only been introduced recently and are therefore not implemented in most browsers (https://caniuse.com/#feat=resizeobserver), altough they are in most recent desktop browsers. I will think about that.

But that will require some more work and thorough testing, so I can't promise that this will happen in the near future.

**Comment by Stumblinbear** [comment]
I'll look into fixes when I can, as this system is going to be released to my asset developers fairly soon, and the lag issue need to be at least somewhat resolved. But, thanks for the info!

**Comment by newcat** [comment]
You can change [this](https://github.com/newcat/baklavajs/blob/master/packages/baklavajs-plugin-renderer-vue/src/components/connection/ConnectionWrapper.vue#L34) line to:
```ts
@Watch("connection.from.parent.position", { deep: true })
@Watch("connection.to.parent.position", { deep: true })
```
which will only watch for position changes of node.

I have never actually worked with resize observers, so the best tip I can give you is to have a look at the [documentation](https://developer.mozilla.org/en-US/docs/Web/API/ResizeObserver). You probably need to subscribe using Vue's `mounted` lifecycle hook and unsubscribe on `beforeDestroy`.

**Comment by Stumblinbear** [comment]
Question.

> This can be avoided by only watching the position properties of the connected nodes; however, this means that height changes of the node are not detected.

If deep watching a connection makes it work correctly, that implies that there's a property that can be watched to detect a size change. Or am I misunderstanding?

**Comment by newcat** [comment]
Yes, but not only one. Vue is watching every value of every interface and option for example (deep in this case really means deep).

So these are the things that might influence the size of a node:
- Connection status of interfaces
- Values of interfaces
- Values of options

If you watch all of them, then yes, it could be done with watchers. But also remember that nodes can dynamically add/remove interfaces or options, so the object containing all interfaces and the object containing all options need to be watched as well. And I probably forgot a few things here as well. That's why I thought it might be easier by using a `ResizeObserver`

**Comment by Stumblinbear** [comment]
Aha! I understand. I was going to suggest propagating changes to connected nodes, but you're right that just watching for size changes would be plenty enough (except maybe a few edge cases, such as removing then adding a new interface in a single update). Perhaps there's a polyfill for ResizeWatcher?

**Comment by newcat** [comment]
Just found a polyfill: https://github.com/juggle/resize-observer
I will try that and - if it does increase performance - implement it.

**Comment by newcat** [comment]
ResizeObservers implemented with c8be5838673242ae0ca0ff8605bd5db6e00ce651 (currently in only in the `@next` version - I will make some more breaking changes soon and then release a 2.0 version. Hope this is fine for you)
